orchagent/intfsorch.cpp

```diff
@@ -193,19 +202,348 @@ bool IntfsOrch::setIntf(const string& alias, sai_object_id_t vrf_id, const IpPre
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * Each newly create interface requires the insertion of two routes in the
+     * system: a subnet route and an ip2me one.
+     */
+    IntfRouteEntry ifSubnetRoute(ifRoute.prefix.getSubnet(),
+                                 ifRoute.ifName,
+                                 "subnet");
+    IntfRouteEntry ifIp2meRoute(getIp2mePrefix(ifRoute.prefix),
+                                ifRoute.ifName,
+                                "ip2me");
+
+    /*
+     * There are two scenarios in which we want to skip the addition of an
+     * interface-subnet route:
+     *
+     * - When dealing with a full-mask interface address (i.e /32 or /128)
+     * - When the port associated to the interface is declared as LOOPBACK
+     */
+    bool subnetOverlap = false, ip2meOverlap = false, skipSubnet = false;
+
+    if (ifSubnetRoute == ifIp2meRoute || port.m_type == Port::LOOPBACK)
+    {
+        skipSubnet = true;
+    }
+
+    if (!skipSubnet)
+    {
+        subnetOverlap = trackIntfRouteOverlap(ifSubnetRoute);
+    }
+    ip2meOverlap = trackIntfRouteOverlap(ifIp2meRoute);
+
+    /* Based on above results, proceed to create routes identified as unique. */
+    if (subnetOverlap)
+    {
+        if (!ip2meOverlap)
+        {
+            addIp2MeRoute(port.m_vr_id, ifIp2meRoute.prefix);
+        }
+    }
+    else /* !subnetOverlap */
+    {
+        if (!skipSubnet)
+        {
+            if (!ip2meOverlap)
+            {
+                addSubnetRoute(port, ifSubnetRoute.prefix);
+                addIp2MeRoute(port.m_vr_id, ifIp2meRoute.prefix);
+            }
+            else
+            {
+                addSubnetRoute(port, ifSubnetRoute.prefix);
+            }
+        }
+        else
+        {
+            if (!ip2meOverlap)
+            {
+                addIp2MeRoute(port.m_vr_id, ifIp2meRoute.prefix);
+            }
+        }
+    }
+
+    /*
+     * A directed-broadcast route is expected in vlan-ipv4 scenarios where the
+     * subnet-length of the associated interface-address is shorter than 30 bits.
+     * If these conditions are met, and there's no overlap with an existing
+     * interface, proceed to create a bcast route.
+     */
+    if (port.m_type == Port::VLAN &&
+        ifRoute.prefix.isV4() &&
+        ifRoute.prefix.getMaskLength() <= 30)
+    {
+        IntfRouteEntry ifBcastRoute(getBcastPrefix(ifRoute.prefix),
+                                    ifRoute.ifName,
+                                    "bcast");
+
+        bool bcastOverlap = trackIntfRouteOverlap(ifBcastRoute);
+        if (!bcastOverlap)
+        {
+            addDirectedBroadcast(port, ifBcastRoute.prefix);
+        }
+    }
+}
+
+/*
+ * Method's goal is to track/record any potential overlap between the interfaces
+ * configured in the system, and alert caller of such an incident.
+ */
+bool IntfsOrch::trackIntfRouteOverlap(const IntfRouteEntry &ifRoute)
+{
```

Comment:
If interfaces are in different VRF,  overlapping addresses should be allowed?

Reply:
Good point. Do you know if we are already supporting that as part of our new VRF implementation?

Reply:
Yes,  vrforch has the VRF implementation with multiple virtual routers.